flink-runtime/src/test/java/org/apache/flink/runtime/rest/compatibility/RestAPIStabilityTest.java

```diff
@@ -0,0 +1,202 @@
+			final List<Tuple2<JsonNode, CompatibilityCheckResult>> compatibilityCheckResults = cur.calls.stream()
+				.map(curCall -> Tuple2.of(curCall, checkCompatibility(oldCall, curCall)))
+				.collect(Collectors.toList());
+
+			if (compatibilityCheckResults.stream().allMatch(result -> result.f1.getBackwardCompatibility() == Compatibility.INCOMPATIBLE)) {
+				fail(oldCall, compatibilityCheckResults);
+			}
+
+			if (compatibilityCheckResults.stream().noneMatch(result -> result.f1.getBackwardCompatibility() == Compatibility.IDENTICAL)) {
+				overallCompatibility = Compatibility.COMPATIBLE;
+			}
+		}
+
+		if (overallCompatibility == Compatibility.COMPATIBLE) {
+			Assert.fail("The API was modified in a compatible way, but the snapshot was not updated. " +
+				"To update the snapshot, re-run this test with -D" + REGENERATE_SNAPSHOT_PROPERTY + " being set.");
+		}
+	}
+
+	private static void fail(final JsonNode oldCall, final List<Tuple2<JsonNode, CompatibilityCheckResult>> compatibilityCheckResults) {
+		final StringBuilder sb = new StringBuilder(8);
+		sb.append("No compatible call could be found for " + oldCall + '.');
+		compatibilityCheckResults.stream()
+			.sorted(Collections.reverseOrder(Comparator.comparingInt(tuple -> tuple.f1.getBackwardCompatibilityGrade())))
+			.forEachOrdered(result -> {
+				sb.append(System.lineSeparator());
+				sb.append("\tRejected by candidate: " + result.f0 + '.');
+
+				sb.append(System.lineSeparator());
+				sb.append("\tCompatibility grade: " + result.f1.getBackwardCompatibilityGrade() + '/' + CompatibilityRoutines.ROUTINES.size());
+
+				sb.append(System.lineSeparator());
+				sb.append("\tIncompatibilities: ");
+
+				for (AssertionError error : result.f1.getBackwardCompatibilityErrors()) {
+					sb.append(System.lineSeparator());
+					sb.append("\t\t" + error.getMessage());
+				}
+			});
+		Assert.fail(sb.toString());
+	}
+
+	private static CompatibilityCheckResult checkCompatibility(final JsonNode oldCall, final JsonNode newCall) {
+		return CompatibilityRoutines.ROUTINES.stream()
+			.map(routine -> checkCompatibility(routine, oldCall, newCall))
+			.reduce(CompatibilityCheckResult::merge)
+			.get();
+	}
+
+	private static <X> CompatibilityCheckResult checkCompatibility(final CompatibilityRoutine<X> routine, final JsonNode oldCall, final JsonNode curCall) {
+		final Optional<X> old = readJson(routine, oldCall);
+		final Optional<X> cur = readJson(routine, curCall);
+
+		return routine.checkCompatibility(old, cur);
+	}
+
+	private static <X> Optional<X> readJson(final CompatibilityRoutine<X> routine, final JsonNode call) {
+		final JsonNode jsonContainer = call.get(routine.getKey());
+
+		try {
+			return jsonContainer != null
+				? Optional.of(OBJECT_MAPPER.treeToValue(jsonContainer, routine.getContainerClass()))
+				: Optional.empty();
+		} catch (JsonProcessingException jpe) {
+			throw new AssertionError(jpe);
+		}
+	}
+
+	static final class RestAPISnapshot {
+		public List<JsonNode> calls;
```

Comment:
The approach of deserializing late is awkward but I do not know a solution yet.

Reply:
The only solution i see is have a well-defined RestAPISnapshot (i.e. pull the details from the routines in here), but tbh I quite like that every routine is able to freely define it's own container.

Reply:
I still do not know a solution but I think I should express myself better why I think current approach can be improved: 
	- `RestAPISnapshot` looks like domain object but contains raw data that is not deserialized
	- two points of deserialization
	- deserialization mixed with business logic, i.e., in the compatibility check we deserialize

I think it's fine that every routine defines it's own container, and there should be a way not to scatter the deserialization logic. Unfortunately I am on vacation now, and I cannot give you a definite recommendation on how to proceed as I have not thought this problem through.

Reply:
I decided that I can live with deserializing the containers late. I think an alternative could be to deserialize eagerly everything as `Object` or a common interface with a custom jackson deserializer.